### PR TITLE
"The Rust Programming Language," Take II

### PR DESCRIPTION
**Pull request by steveklabnik**
An updated version of https://github.com/rust-lang/rust/pull/19461

This version vendors aturon/rust-book@731f7bf and builds it when building the docs. This is almost great, except my `make`-foo is poor, so I have my own personal paths in `mk/docs.mk`. How should I best get around that?

/cc @brson

### CODE REVIEW

File: src/rustbook/build.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        // write the postlude to a temporary HTML file for rustdoc inclusion
+        let postlude = tmp.path().join("postlude.html");
+        {
+            let mut toc = BufferedWriter::new(try!(File::create(&postlude)));
+            try!(writeln!(&mut toc, "</div></div>"));
+        }
+
+        try!(fs::mkdir_recursive(&out_path, io::USER_DIR));
+
+        let output_result = Command::new("rustdoc")
```
Comment (steveklabnik) [inline src/rustbook/build.rs]: :cry:

File: mk/docs.mk [hunk 2]
```diff
@@ -265,3 +269,11 @@ endif
 
 docs: $(DOC_TARGETS)
 compiler-docs: $(COMPILER_DOC_TARGETS)
+
+RUSTC = $(STAGE2_T_$(CFG_BUILD)_H_$(CFG_BUILD))
+
+trpl: $(RUSTBOOK_EXE) $(wildcard $(S)/src/doc/trpl/*.md)
+	$(Q) rm -Rf tmp/trpl && rm -Rf doc/book
+	$(Q) mkdir tmp/trpl
+	$(Q) $(RUSTBOOK) build $(S)src/doc/trpl tmp/trpl
```
Comment (steveklabnik) [inline mk/docs.mk]: I thought that was part of the out of tree bit.

File: src/rustbook/build.rs (hunk 1 above)
Comment (steveklabnik) [inline src/rustbook/build.rs]: @aturon what do you think?

**Comment by steveklabnik** [comment]
Okay! This is almost there. One error I'm not sure about:

```
rustc: x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/bin/rustbook
cp: x86_64-unknown-linux-gnu/stage2/bin/rustbook
1. Introduction
error opening ` tmp/trpl/README.html` for writing: couldn't create file (no such file or directory (No such file or directory); path= tmp/trpl/README.html; mode=truncate; access=write)
```

any ideas, @brson or @alexcrichton ?

**Comment by brson** [comment]
There appears to be an extra space in the path rustbook is trying to open "<space>tmp/trpl..."

File: src/doc/trpl/Makefile [hunk 3]
```diff
@@ -0,0 +1,4 @@
+all: _book
+	
+_book: SUMMARY.md README.md src
+	rustbook build
```
Comment (steveklabnik) [inline src/doc/trpl/Makefile]: The README is the root page, so it can't, but it should have real text.

File: mk/docs.mk [hunk 4]
```diff
@@ -265,3 +269,11 @@ endif
 
 docs: $(DOC_TARGETS)
 compiler-docs: $(COMPILER_DOC_TARGETS)
+
+RUSTC = $(STAGE2_T_$(CFG_BUILD)_H_$(CFG_BUILD))
+
+trpl: $(RUSTBOOK_EXE) $(wildcard $(S)/src/doc/trpl/*.md)
```
Comment (steveklabnik) [inline mk/docs.mk]: okay, so i pushed a commit that does the tmp/trpl.ok stuff, but I accidentally forgot to `touch mk/docs.mk`, and it's late, so I'm not gonna do an actual re-build right now. 

what's the best way to include it into `make`, do you think? And, I need to make the mirror `test` rule too.

**Comment by steveklabnik** [comment]
This is getting close to being good to go. Left to do:

1. write the makefile rule that calls `rustbook test`
2. integrate `make trpl` and `make trpl-test` into `make docs`.

Anything else?

**Comment by alexcrichton** [comment]
I've rebased this branch on the current master here: https://github.com/alexcrichton/rust/commit/6a5b558b2a7d9bc121dc9a5d3e687ef80d3c3848

In that commit I've also integrated tests into the default build (not via `rustbook test` though), and I've also integrated the docs into the regular build.

Some other agenda items I saw:

* The main landing page needs to have many links updated
* The guides need to be all brought back up to date as a bunch of changes have happened since they were moved.
* A quick pass to make sure everything looks as expected.

Can't wait for this to land!

**Comment by steveklabnik** [comment]
Ah yes. I'm going to eat breakfast, and then I'll tackle that.

**Comment by steveklabnik** [comment]
Okay! This should be good to go, I think. I'd imagine we'd want to sqaush some of this, yeah?

**Comment by steveklabnik** [comment]
2015-01-08-112123_1022x59_scrot

:cry: it is done

File: src/doc/trpl/SUMMARY.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    * [Tasks](src/tasks.md)
+    * [Error Handling](src/error-handling.md)
+* [III: Advanced Topics](src/advanced.md)
+    * [FFI](src/ffi.md)
+    * [Unsafe Code](src/unsafe.md)
+    * [Macros](src/macros.md)
+    * [Compiler Plugins](src/plugins.md)
+* [Conclusion](src/conclusion.md)
+
+vim src/doc/trpl/src/installing-rust.md src/doc/trpl/src/hello-world.md src/doc/trpl/src/hello-cargo.md src/doc/trpl/src/variable-bindings.md src/doc/trpl/src/if.md src/doc/trpl/src/functions.md src/doc/trpl/src/comments.md src/doc/trpl/src/compound-data-types.md src/doc/trpl/src/match.md src/doc/trpl/src/looping.md src/doc/trpl/src/strings.md src/doc/trpl/src/arrays-vectors-and-slices.md src/doc/trpl/src/standard-input.md src/doc/trpl/src/guessing-game.md
```
Comment (alexcrichton) [inline src/doc/trpl/SUMMARY.md]: odd line?

File: src/doc/trpl/src/basic.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+% Basics
+
+Basic Rust stuff.
```
Comment (alexcrichton) [inline src/doc/trpl/src/basic.md]: Just confirming this is intended.

File: src/doc/trpl/src/advanced.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+% Advanced
+
+Advanced Rust stuff.
```
Comment (alexcrichton) [inline src/doc/trpl/src/advanced.md]: Just confirming this is intended.

File: src/doc/trpl/src/closures.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+Rust has a second type of closure, called a **proc**. Procs are created
+with the `proc` keyword:
+
+```{rust}
+let x = 5i;
+
+let p = proc() { x * x };
+println!("{}", p()); // prints 25
+```
```
Comment (alexcrichton) [inline src/doc/trpl/src/closures.md]: Hm, do the tests pass? This looks like a testable example but I would kinda hope we reject this!

Comment (steveklabnik) [inline src/doc/trpl/src/closures.md]: oh no, this makes me question my rebasing skillz

File: src/doc/trpl/SUMMARY.md (hunk 5 above)
Comment (steveklabnik) [inline src/doc/trpl/SUMMARY.md]: lol i suck at vim

File: src/doc/trpl/src/basic.md (hunk 6 above)
Comment (steveklabnik) [inline src/doc/trpl/src/basic.md]: all these addressed

**Comment by alexcrichton** [comment]
Merged in #20760

**Comment by alexcrichton** [comment]
Can't wait to see this on doc.rust-lang.org, really nice work here @steveklabnik!

**Comment by steveklabnik** [comment]
:heart_eyes: :confetti_ball:

**Comment by Gankro** [comment]
Woaaah! Congrats!

File: src/doc/trpl/README.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+This section contains individual chapters, which are self-contained. They focus
+on specific topics, and can be read in any order.
+
+After reading "Intermediate," you will have a solid understanding of Rust,
+and will be able to understand most Rust code and write more complex programs.
+
+## Advanced
+
+In a similar fashion to "Intermediate," this setion is full of individual,
```
Comment (asterite) [inline src/doc/trpl/README.md]: Should be 'section'

Comment (steveklabnik) [inline src/doc/trpl/README.md]: Totally! Please open an issue or PR about this. Thanks :)

Comment (Gankro) [inline src/doc/trpl/README.md]: This PR is just a bulk move of all the docs; the actual contents aren't under review here (feel free to submit a seperate issue or PR).

**Pull request closed without merging**